File: ruby/single_number.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @return {Integer}
def single_number(nums)
  set = Set.new

  nums.each do |num|
    if set.include?(num)
      set.delete(num)
    else
      set.add(num)
    end
  end

  set.first
end

File: python/single_number.py
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        unique_nums = set()

        for num in nums:
            if num in unique_nums:
                unique_nums.remove(num)
            else:
                unique_nums.add(num)

        return unique_nums.pop()
